crates/resolver/src/dns_lru/ttl_config_deserialize.rs

```diff
@@ -0,0 +1,555 @@
+use std::collections::hash_map::Entry;
+
+use serde::{
+    de::{value::StrDeserializer, Error, MapAccess, Visitor},
+    Deserialize,
+};
+
+use hickory_proto::rr::RecordType;
+
+use super::TtlConfig;
+
+const TTL_CONFIG_FIELDS: [&str; 37] = [
+    "default",
+    "A",
+    "AAAA",
+    "ANAME",
+    "ANY",
+    "AXFR",
+    "CAA",
+    "CDS",
+    "CDNSKEY",
+    "CERT",
+    "CNAME",
+    "CSYNC",
+    "DNSKEY",
+    "DS",
+    "HINFO",
+    "HTTPS",
+    "IXFR",
+    "KEY",
+    "MX",
+    "NAPTR",
+    "NS",
+    "NSEC",
+    "NSEC3",
+    "NSEC3PARAM",
+    "NULL",
+    "OPENPGPKEY",
+    "OPT",
+    "PTR",
+    "RRSIG",
+    "SIG",
+    "SOA",
+    "SRV",
+    "SSHFP",
+    "SVCB",
+    "TLSA",
+    "TSIG",
+    "TXT",
+];
+
+// This is implemented manually in order to leverage the existing deserialization code for
+// `RecordType`.
+//
+// A derived implementation cannot be used without requiring an extra level of TOML tables, because
+// `#[serde(flatten)]` and `#[serde(deny_unknown_fields)]` are incompatible.
+impl<'de> Deserialize<'de> for TtlConfig {
```

Comment:
Similar question here: can we instead define a private type such that we can use `serde(from)` or `serde(try_from)` instead of all this stuff?

Reply:
I figured something out using `serde(from)` and some more attributes on the field enum. The error messages are a bit worse, but on par with what I expect from serde.